felix/bpf/maps.go

```diff
@@ -314,15 +349,79 @@ func (b *PinnedMap) Open() error {
 	return err
 }
 
+// nolint
+func (b *PinnedMap) pinnedMapMatchesConfiguration(maxEntries int) bool {
+	return maxEntries == b.MaxEntries
+}
+
+// nolint
+func (b *PinnedMap) migratePinnedMap(from, to string) error {
+	err := RepinMap(b.versionedName(), to)
+	if err != nil {
+		return fmt.Errorf("error repinning %s to %s: %w", from, to, err)
+	}
+	err = os.Remove(from)
+	if err != nil {
+		return fmt.Errorf("error removing the pin %s", from)
+	}
+	return nil
+}
+
+func (b *PinnedMap) oldMapExists() bool {
+	_, err := os.Stat(b.Path() + "_old")
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false
+		}
+	}
+	return true
+}
+
 func (b *PinnedMap) EnsureExists() error {
+	var oldfd MapFD
+	oldMapPath := b.Path() + "_old"
 	if b.fdLoaded {
 		return nil
 	}
 
+	// In case felix restarts in the middle of migration, we might end up with
+	// old map. Repin the old map and let the map creation continue.
+	if b.oldMapExists() {
+		if _, err := os.Stat(b.Path()); err == nil {
+			os.Remove(b.Path())
+		}
+		err := b.migratePinnedMap(oldMapPath, b.Path())
+		if err != nil {
+			return fmt.Errorf("error repinning old map %s to %s, err=%w", oldMapPath, b.Path(), err)
+		}
+	}
+
 	if err := b.Open(); err == nil {
-		return nil
+		// store the old fd
+		oldfd = b.MapFD()
+
+		// Get the existing map info
+		mapInfo, err := GetMapInfo(b.fd)
+		if err != nil {
+			return fmt.Errorf("error getting map info of the pinned map %w", err)
+		}
+
+		if b.pinnedMapMatchesConfiguration(mapInfo.MaxEntries) {
+			return nil
+		}
+		err = b.migratePinnedMap(b.Path(), oldMapPath)
+		if err != nil {
+			return fmt.Errorf("error migrating the old map %w", err)
+		}
 	}
 
+	// Close the oldfd for the maps to be removed from the kernel.
+	defer func() {
+		if oldfd != 0 {
+			oldfd.Close()
+		}
+	}()
```

Comment:
I think this should be like

```
	if err := b.Open(); err == nil {
		oldfd = b.MapFD()
                defer oldfd.Close()
```

Reply:
then do the defer after you returned nil in case the map had the same size. 0 is a unlikely but valid FD. But Now as the code has reshuffled, `oldfd` is not used here anymore so it should be closed above without a need for defer.